File: packages/wasmvm/wasmlib/go/wasmlib/wasmtypes/scuint64.go
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

package wasmtypes

import (
	"encoding/binary"
	"strconv"
)

const ScUint64Length = 8

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

func Uint64Decode(dec *WasmDecoder) uint64 {
	return dec.VluDecode(64)
}

func Uint64Encode(enc *WasmEncoder, value uint64) {
	enc.VluEncode(value)
}

func Uint64FromBytes(buf []byte) uint64 {
	if len(buf) == 0 {
		return 0
	}
	if len(buf) != ScUint64Length {
		panic("invalid Uint64 length")
	}
	return binary.LittleEndian.Uint64(buf)
}

func Uint64ToBytes(value uint64) []byte {
	tmp := make([]byte, ScUint64Length)
	binary.LittleEndian.PutUint64(tmp, value)
	return tmp
}

func Uint64ToString(value uint64) string {
	return strconv.FormatUint(value, 10)
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScImmutableUint64 struct {
	proxy Proxy
}

func NewScImmutableUint64(proxy Proxy) ScImmutableUint64 {
	return ScImmutableUint64{proxy: proxy}
}

func (o ScImmutableUint64) Exists() bool {
	return o.proxy.Exists()
}

func (o ScImmutableUint64) String() string {
	return Uint64ToString(o.Value())
}

func (o ScImmutableUint64) Value() uint64 {
	return Uint64FromBytes(o.proxy.Get())
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

type ScMutableUint64 struct {
	ScImmutableUint64
}

func NewScMutableUint64(proxy Proxy) ScMutableUint64 {
	return ScMutableUint64{ScImmutableUint64{proxy: proxy}}
}

func (o ScMutableUint64) Delete() {
	o.proxy.Delete()
}

func (o ScMutableUint64) SetValue(value uint64) {
	o.proxy.Set(Uint64ToBytes(value))
}

File: packages/wasmvm/wasmlib/ts/wasmlib/wasmtypes/scuint64.ts
// Copyright 2020 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {panic} from '../sandbox';
import {uintFromString64, WasmDecoder, WasmEncoder} from './codec';
import {Proxy} from './proxy';
import {uint32FromBytes} from './scuint32';

export const ScUint64Length = 8;

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export function uint64Decode(dec: WasmDecoder): u64 {
    return dec.vluDecode64(64);
}

export function uint64Encode(enc: WasmEncoder, value: u64): void {
    enc.vluEncode64(value);
}

export function uint64FromBytes(buf: Uint8Array): u64 {
    if (buf.length == 0) {
        return 0n;
    }
    if (buf.length != ScUint64Length) {
        panic('invalid Uint64 length');
    }
    // let ret: u64 = buf[7];
    // ret = (ret << 8) | buf[6];
    // ret = (ret << 8) | buf[5];
    // ret = (ret << 8) | buf[4];
    // ret = (ret << 8) | buf[3];
    // ret = (ret << 8) | buf[2];
    // ret = (ret << 8) | buf[1];
    // return (ret << 8) | buf[0];
    const u32high = uint32FromBytes(buf.subarray(4));
    const u32low = uint32FromBytes(buf.subarray(0, 4));
    return BigInt(u32high) * 0x100000000n + BigInt(u32low);
}

export function uint64ToBytes(value: u64): Uint8Array {
    const buf = new Uint8Array(ScUint64Length);
    const u32low = Number(value % 0x100000000n);
    const u32High = Number(value / 0x100000000n % 0x100000000n);
    buf[0] = u32low as u8;
    buf[1] = (u32low >> 8) as u8;
    buf[2] = (u32low >> 16) as u8;
    buf[3] = (u32low >> 24) as u8;
    buf[4] = u32High as u8;
    buf[5] = (u32High >> 8) as u8;
    buf[6] = (u32High >> 16) as u8;
    buf[7] = (u32High >> 24) as u8;
    return buf;
}

export function uint64FromString(value: string): u64 {
    return uintFromString64(value, 64);
}

export function uint64ToString(value: u64): string {
    return value.toString();
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScImmutableUint64 {
    proxy: Proxy;

    constructor(proxy: Proxy) {
        this.proxy = proxy;
    }

    exists(): bool {
        return this.proxy.exists();
    }

    toString(): string {
        return uint64ToString(this.value());
    }

    value(): u64 {
        return uint64FromBytes(this.proxy.get());
    }
}

// \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\ // \\

export class ScMutableUint64 extends ScImmutableUint64 {
    delete(): void {
        this.proxy.delete();
    }

    setValue(value: u64): void {
        this.proxy.set(uint64ToBytes(value));
    }
}
